watertap/tests/test_notebooks.py

```diff
@@ -0,0 +1,98 @@
+"""
+Test Jupyter notebooks in the docs.
+"""
+# stdlib
+import logging
+from pathlib import Path
+import time
+
+# deps
+import nbformat
+from nbconvert.preprocessors import ExecutePreprocessor, CellExecutionError
+import pytest
+
+# package
+from watertap.edb import ElectrolyteDB
+
+# Logging setup
+
+_log = logging.getLogger(__name__)
+
+# Constants
+
+JUPYTER_NB_VERSION = 4
+EXECUTE_TIMEOUT_SEC = 600
+
+# Exceptions
+
+
+class NotebookError(Exception):
+    pass
+
+
+# Utility
+# =======
+
+
+def run_notebook(path: Path):
+    # parse
+    _log.debug(f"begin: parsing '{path}'")
+    with path.open("r", encoding="utf-8") as f:
+        nb = nbformat.read(f, as_version=JUPYTER_NB_VERSION)
+    _log.debug(f"end: parsing '{path}'")
+
+    # execute
+    pp = ExecutePreprocessor(timeout=EXECUTE_TIMEOUT_SEC)
+    _log.debug(f"begin: executing '{path}'")
+    t0 = time.time()
+    try:
+        metadata = {"metadata": {"path": path.parent}}
+        pp.preprocess(nb, metadata)
+    except (CellExecutionError, NameError) as err:
+        _log.warning(f"Failed: {path} :: {err}")
+        raise NotebookError(f"execution error for '{path}': {err}")
+    except TimeoutError as err:
+        _log.warning(f"Timeout: {path} :: {err}")
+        dur, timeout = time.time() - t0, EXECUTE_TIMEOUT_SEC
+        raise NotebookError(f"timeout for '{path}': {dur}s > {timeout}s")
+    dur = time.time() - t0
+    _log.debug(f"end: executing '{path}' duration={dur}s")
+
+
+def find_notebooks(path: Path):
+    return path.glob("**/*.ipynb")
+
+
+@pytest.fixture(scope="session")
+def docs_root(dirname="docs"):
+    p = Path(__file__).parent
+    while not (p / dirname).exists():
+        pp = p.parent
+        if pp == p:
+            raise RuntimeError(f"Could not find '{dirname}' dir")
+        p = pp
+    yield p / dirname
+
+
+def mongodb():
+    try:
+        edb = ElectrolyteDB()  # will fail if no DB
+        edb.get_base()  # will fail if DB is not loaded
+    except Exception as err:
+        _log.warning(f"Could not connect to MongoDB: {err}")
+    return False
+
+# Tests
+# =====
+
+
+@pytest.mark.skipif(not mongodb(), reason="MongoDB is required")
+@pytest.mark.component
+def test_edb_notebooks(docs_root):
```

Comment:
This works, but the check is performed at import/collection time.

One alternative to perform the check at runtime instead is to turn this into a fixture:

```suggestion
@pytest.fixture(scope="session")
def mongodb_check():
    try:
        edb = ElectrolyteDB()  # will fail if no DB
        edb.get_base()  # will fail if DB is not loaded
    except Exception as err:
        pytest.skip(f"Could not connect to MongoDB: {err}")

# Tests
# =====


@pytest.mark.usefixtures('mongodb_check')
@pytest.mark.component
def test_edb_notebooks(docs_root):
```

Reply:
If I don't perform this at the import time, then I can't tell whether the test was skipped or not. I prefer it this way, in the case where a real DB is needed. We can assure it runs once if necessary.

Reply:
> If I don't perform this at the import time, then I can't tell whether the test was skipped or not

Wait, I'm not sure I understand. What do you mean with this? I think in general performing the DB connection at runtime (i.e. using `pytest.skip()` from within the fixture function body) rather than at collect time is preferable, but maybe I'm missing some drawbacks of this approach.

Reply:
sorry I didn't actually read your code. No, that would work, essentially doing the skip test later.